Interactive Tutorial on QDMI Device Implementation (#330)
Munich-Quantum-Software-Stack/QDMI: docs/tutorial.md

# Tutorial: Implementing a QDMI Device

This guide walks you through implementing a minimal **QDMI device**. By the end of this tutorial, you'll have a functional implementation capable of handling simulated quantum workloads. We've designed this as an interactive walkthrough where you build and verify features incrementally.

The **Quantum Device Management Interface (QDMI)** is a standardized layer for hardware abstraction. We're building a bridge that allows high-level drivers to control your device implementation through a stable C interface.

```mermaid
graph LR
    accTitle: QDMI Architecture Overview
    accDescr: High-level diagram showing data flow from a Driver through the QDMI API to your Device Implementation and finally to the hardware.

    Driver[Driver / User App] -->|Request| API[QDMI API]
    API -->|Invoke| Device[Your Device Implementation]
    Device <-->|Control| HW[Quantum Hardware / Simulator]

> ```

Reviewer: I like these diagrams a lot. Have you tested whether they get rendered successfully in the docs by building the target `qdmi_docs` and then opening the HTML file located in `html/...`?

With respect to this particular diagram, for a newbie I find "Driver / User App" confusing and would just replace it with "User" for the sake of simplicity. I think this option is better than actually modelling the Driver plus multiple Clients in this diagram.

Additionally, I would like to have the arrows and their labelling such that they indicate that the information flow is also all the way back from the quantum hardware to the user.

### Your Progress Journey

- [ ] **Phase 1**: [Project Creation](#tutorial-create)
- [ ] **Phase 2**: [Global Lifecycle](#tutorial-global)
- [ ] **Phase 3**: [Session Handling](#tutorial-session)
- [ ] **Phase 4**: [Device Properties](#tutorial-query)
- [ ] **Phase 5**: [Job Execution](#tutorial-jobs)

## Prerequisites {#tutorial-prerequisites}

The template requires a few standard development tools:

- **C++ Compiler**: Supporting C++20 (GCC 10+, Clang 10+, or MSVC 19.29+).

> - **C Compiler**: Supporting C11.

Reviewer: Can we mention here, that this requirement is covered by the recommended options above?

- **CMake**: Version 3.24 or higher.

> - **Python**: Version 3.10 or higher.

Reviewer: We usually use `uv` for Python installations. Could you add a recommandation for that?

> CMake handles fetching all other dependencies, such as the QDMI core interface and the GoogleTest framework, during the configuration step.

Reviewer: Not sure, but should we add a node that the CMake configuration step requires a internet connection for exactly this purpose?

## Phase 1 — Project Setup {#tutorial-create}

> First, we'll generate the project workspace using the QDMI template. For this guide, we'll use `tutorial` as our project prefix.

Reviewer: ```suggestion
First, we'll generate the project workspace using the QDMI template. This requires setting a prefix that is used within the template. For this guide, we'll use `tutorial` as our project prefix.
```
Reads better imho.

> ```sh

Reviewer: Do you want to add a step for cloning the git repository, i.e., `git clone...`?

# Step 1: Generate the project files
cmake -DQDMI_GENERATE_TEMPLATE=ON \
      -DTEMPLATE_PREFIX="tutorial" \

>       -DTEMPLATE_PATH="my_qdmi_device" \

Reviewer: ```suggestion
      -DTEMPLATE_PATH="tutorial_device" \
```
I would prefer a matching name with the prefix here.

      -S . -B build

# Step 2: Actually write the files to disk
cmake --build build --target qdmi-template

# Step 3: Enter the new project directory
cd my_qdmi_device
```

The resulting directory (`my_qdmi_device`) follows a standard structure:

```text
.
├── src/                # Your implementation (tutorial_device.cpp)
├── test/               # Your tests (test_tutorial_device.cpp)
├── cmake/              # Build logic
├── CMakeLists.txt      # Main project file
└── pyproject.toml      # Python configuration
```

### The Verification Loop

We use a simple verification loop to track progress: you'll set up the tests first, confirm they fail, implement the fix, and then verify the results.

> **Action Required**: Copy the full test suite from the [Test Suite Reference](#tutorial-test-reference) section at the bottom of this page and paste it into `test/test_tutorial_device.cpp` now.

Reviewer: While this approach works, I was thinking whether one could integrate the test suite into the template in the following way. When one instatiates the template in the same way as you described above, nothing changes, the exact files are written. However, when one set a new CMake variable `TEMPLATE_TEST_SUITE:BOOL=ON` then the test suite gets automatically copied into the respective directory of the device. Do you think this is possible?

Building the tests at this stage should result in failures, which is expected as we haven't implemented the logic yet.

```sh
# Build the test target
cmake -S . -B build -DCMAKE_BUILD_TYPE=Release
cmake --build build --target tutorial-qdmi-device-test

# Run the tests
ctest --test-dir build -C Release
```

> [!TIP]
> **Progress Check**: Starting with failures is expected—we'll address these one by one as we implement the required functions.

> [!NOTE]
> **Before writing any implementation code**, ensure `src/tutorial_device.cpp`
> starts with these includes. They provide the core interface, string handling,
> and memory utilities used throughout the implementation:
>
> ```cpp
> #include "tutorial_qdmi/device.h"
> #include <string>    // For std::string
> #include <cstring>   // For std::memcpy
> #include <stdexcept> // For std::bad_alloc

> > ```

Reviewer: I would put that into phase 2 as this is the first phase where one starts working with the `.cpp` file.

## Phase 2 — Global Device Lifecycle {#tutorial-global}

The driver interacts with your device starting with `initialize` and ending with `finalize`. In a real implementation, this is where you'd typically connect to physical hardware or allocate runtime resources; for this tutorial, we'll keep the logic minimal.

Open `src/tutorial_device.cpp` and update these stubs to return `QDMI_SUCCESS`.

```cpp
int tutorial_QDMI_device_initialize() {
  // Global startup logic (e.g., hardware connection) goes here
  return QDMI_SUCCESS;
}

int tutorial_QDMI_device_finalize() {
  // Global cleanup logic goes here
  return QDMI_SUCCESS;
}
```

> [!TIP]
> **Check Now**:
> Run the tests again. **Checkpoint 0** should now pass.

> > `ctest --test-dir build -C Release -R "Init"`

Reviewer: I am a big fan of consistency. Could one either use "Checkpoint 0" also in the command instead of `"Init"` or vice versa, i.e., call the checkpoint "Checkpoint Init", whatever you prefer?

## Phase 3 — Session Handling {#tutorial-session}

> Once the global interface is ready, we need a way to manage connections. This is handled through a **Session**.

Reviewer: While reading through the file here, it would be cool if concepts like "Session" here would link to the respective pages in the existing documentation. Since the documentation is rendered using `doxygen` this is as easy as writing `@ref QDMI_Device_Session "Session"` or `\ref QDMI_Device_Session "Session"`. You can also omit the custom label, i.e., `"Session", the the name of the concept will be displayed.

In QDMI, sessions are managed via **handles**. A handle is essentially an opaque pointer to a struct that your implementation defines. Specifically, `tutorial_QDMI_Device_Session` is a typedef from the C API, while the internal struct (`tutorial_QDMI_Device_Session_impl_d`) is a private implementation detail.

```mermaid
stateDiagram-v2
    accTitle: Session State Lifecycle
    accDescr: State diagram showing transitions from Allocation to Initialization, and finally to being Freed.

    [*] --> Allocated: session_alloc()
    Allocated --> Allocated: set_parameter(TOKEN)
    Allocated --> Initialized: session_init()
    Initialized --> [*]: session_free()

> ```

Reviewer: This is cool. I would like to have two things changed:
1. One can set more than a token, at least in the future. Could one frame that in a way that it is clear that this is more general than just for setting a token?
2. After the session is initialised it can be used for querying information. This should also be covered somehow in the diagram as currently establishing a session seems pointless when one only looks at the diagram.

Define a basic struct in `src/tutorial_device.cpp` to track the session state.

> [!NOTE]
> Using a private C++ struct to hold session state helps keep the public C interface stable and avoids ABI versioning headaches between your implementation and the driver.

```cpp
enum class SESSION_STATUS { ALLOCATED, INITIALIZED };

struct tutorial_QDMI_Device_Session_impl_d {
  std::string token;
  SESSION_STATUS status = SESSION_STATUS::ALLOCATED;
};
```

### 2. Implement the Logic

Implement the session management functions. We include safety checks here to ensure the implementation is robust against null handles.

```cpp
int tutorial_QDMI_device_session_alloc(tutorial_QDMI_Device_Session *session) {
  if (session == nullptr) return QDMI_ERROR_INVALIDARGUMENT;

  try {
    *session = new tutorial_QDMI_Device_Session_impl_d();
    return QDMI_SUCCESS;
  } catch (const std::bad_alloc&) {
    return QDMI_ERROR_OUTOFMEM;
  }
}

int tutorial_QDMI_device_session_set_parameter(tutorial_QDMI_Device_Session session,
                                            QDMI_Device_Session_Parameter param,
                                            const size_t size, const void *value) {

>   if (session == nullptr) return QDMI_ERROR_INVALIDARGUMENT;

Reviewer: `INVALIDARGUMENT` should also be returned if the value is not `NULL` but `size == 0`. Could you please check the docs for all error cases and cover them in this tutorial implementation?

Only supporting the token parameter is a very good choice.

Reviewer: This would be also something that should be covered in the tests, i.e., that the return values, in particular the error codes are in line with the specification. Here, you can build on the existing testing infrastructure. However, please check, because not all cases are always covered.

  if (session->token.empty()) return QDMI_ERROR_PERMISSIONDENIED;

  session->status = SESSION_STATUS::INITIALIZED;
  return QDMI_SUCCESS;
}

void tutorial_QDMI_device_session_free(tutorial_QDMI_Device_Session session) {
  if (session != nullptr) delete session;
}
```

> [!TIP]
> **Check Now**:
> With the handshake logic in place, you can now verify session handling. **Checkpoints 1 and 2** should pass.
> `ctest --test-dir build -C Release -R "Alloc|Init"`

## Phase 4 — First Query (Device Properties) {#tutorial-query}

The driver uses properties to retrieve information about your device, such as its name or qubit count. Since this data can vary in length, we use a **Two-Step Query Pattern**.

```mermaid
sequenceDiagram
    accTitle: Two-Step Query Handshake
    accDescr: Sequence diagram illustrating the handshake between the Driver and the Device to query a property.
    participant D as Driver
    participant V as Device (You)
    D->>V: query_property(prop, value=NULL)
    Note over V: Calculate required size
    V-->>D: QDMI_SUCCESS, size_ret=N
    D->>V: query_property(prop, value=Buffer[N])
    Note over V: Copy data to buffer
    V-->>D: QDMI_SUCCESS

> ```

Reviewer: Depending on the modifications above, consider renaming the Driver to User even though Driver is absolutely correct, the use of the term User may result in less confusion.

> [!NOTE]
> While we use two steps for safety, QDMI also allows retrieving data in a single call if the driver already knows the correct size or provides a sufficiently large buffer.

### Implementing the Queries

Define the device constants and implement the query function in `src/tutorial_device.cpp`.

> [!NOTE]
> This switch-based pattern is standard in QDMI. It looks a bit verbose, but it's efficient and prevents the driver from needing to know your implementation's internal memory layout.

```cpp
const std::string DEVICE_NAME = "MyTutorialDevice";
const size_t QUBIT_COUNT = 1;

int tutorial_QDMI_device_session_query_device_property(
    tutorial_QDMI_Device_Session session, const QDMI_Device_Property prop,
    const size_t size, void *value, size_t *size_ret) {

  if (session == nullptr) return QDMI_ERROR_INVALIDARGUMENT;

>   if (session->status != SESSION_STATUS::INITIALIZED) return QDMI_ERROR_BADSTATE;
> 
>   switch (prop) {
>     case QDMI_DEVICE_PROPERTY_NAME: {
>       size_t name_size = DEVICE_NAME.size() + 1;
>       if (size_ret) *size_ret = name_size;
>       if (value == nullptr) return QDMI_SUCCESS; // Step 1: size request
> 
>       if (size < name_size) return QDMI_ERROR_INVALIDARGUMENT;
>       std::memcpy(value, DEVICE_NAME.c_str(), name_size);
>       return QDMI_SUCCESS;
>     }

Reviewer: ```suggestion
      // `+1` to also include the `\0` at the end of the string
      size_t name_size = DEVICE_NAME.size() + 1;
      if (size_ret) {
        *size_ret = name_size;
      }
      if (value != nullptr) {
        if (size < name_size) {
          return QDMI_ERROR_INVALIDARGUMENT;
        }
        std::memcpy(value, DEVICE_NAME.c_str(), name_size);
      }
      return QDMI_SUCCESS;
    }
```
Also appleis to the following code.

    case QDMI_DEVICE_PROPERTY_SITES: {
      if (size_ret) *size_ret = sizeof(size_t);
      if (value == nullptr) return QDMI_SUCCESS;

      *static_cast<size_t*>(value) = QUBIT_COUNT;
      return QDMI_SUCCESS;
    }
    default: return QDMI_ERROR_NOTSUPPORTED;
  }
}

> ```

Reviewer: In the example device we used macros to make the definition of properties easier and to reduce the amount of boilerplate code. While I prefer the implementation you provided for the tutorial, I would like to have a note that mentions the other approach with a link to the resources.

> [!TIP]
> **Check Now**:
> The device can now describe itself to the driver. **Checkpoint 3** should pass.

> > `ctest --test-dir build -C Release -R "Query"`

Reviewer: Is there a reason why this checkpoint does not cover the previous tests anymore?

## Phase 5 — Job Handling {#tutorial-jobs}

Quantum programs are managed via **Jobs**. Jobs follow a specific lifecycle, moving from creation to submission and eventually to completion.

```mermaid
stateDiagram-v2
    accTitle: Job Handling Lifecycle
    accDescr: State diagram showing a job's journey from creation to submission and completion.
    [*] --> Created: create_device_job()
    Created --> Submitted: job_submit()
    Submitted --> Done: Simulation
    Done --> [*]: job_free()

> ```

Reviewer: While I understand that the result retrieval is not part of the diagram for simplicity, I would like to see a placeholder representing the result retrieval in the diagram. Otherwise, the creation of a job seem spointless.

};
```

### 2. Implement Job Logic

The following functions handle creating and executing jobs. In this tutorial, we simulate immediate successful completion.

```cpp
int tutorial_QDMI_device_session_create_device_job(tutorial_QDMI_Device_Session session,
                                                tutorial_QDMI_Device_Job *job) {
  if (session == nullptr || job == nullptr) return QDMI_ERROR_INVALIDARGUMENT;
  try {
    *job = new tutorial_QDMI_Device_Job_impl_d();
    return QDMI_SUCCESS;
  } catch (const std::bad_alloc&) {
    return QDMI_ERROR_OUTOFMEM;
  }
}

int tutorial_QDMI_device_job_set_parameter(tutorial_QDMI_Device_Job job,
                                        const QDMI_Device_Job_Parameter param,
                                        const size_t size, const void *value) {
  if (job == nullptr || value == nullptr) return QDMI_ERROR_INVALIDARGUMENT;

  if (param == QDMI_DEVICE_JOB_PARAMETER_PROGRAM) {
    job->program = std::string(static_cast<const char*>(value), size);
    return QDMI_SUCCESS;
  }
  return QDMI_ERROR_NOTSUPPORTED;
}

int tutorial_QDMI_device_job_submit(tutorial_QDMI_Device_Job job) {
  if (job == nullptr) return QDMI_ERROR_INVALIDARGUMENT;

>   if (job->program.empty()) return QDMI_ERROR_BADSTATE;

Reviewer: Also if the job has already been submitted earlier.

  void SetUp() override {
    ASSERT_EQ(tutorial_QDMI_device_initialize(), QDMI_SUCCESS)
        << "Checkpoint 0 Failed: Basic device initialization returned an error.";
  }
  void TearDown() override { tutorial_QDMI_device_finalize(); }
};

class QDMISessionTest : public QDMIBaseTest {
protected:
  tutorial_QDMI_Device_Session session = nullptr;
  void SetUp() override {
    QDMIBaseTest::SetUp();
    ASSERT_EQ(tutorial_QDMI_device_session_alloc(&session), QDMI_SUCCESS)
        << "Checkpoint 1 Failed: Could not allocate a session handle.";
    const std::string token = "tutorial_token";
    ASSERT_EQ(tutorial_QDMI_device_session_set_parameter(session, QDMI_DEVICE_SESSION_PARAMETER_TOKEN, token.size(), token.c_str()), QDMI_SUCCESS);
    ASSERT_EQ(tutorial_QDMI_device_session_init(session), QDMI_SUCCESS)
        << "Checkpoint 2 Failed: Session initialization failed.";
  }
  void TearDown() override {
    if (session) tutorial_QDMI_device_session_free(session);
    QDMIBaseTest::TearDown();
  }
};

TEST_F(QDMIBaseTest, Initialization) {
  // Checkpoint 0: Verified by SetUp/TearDown
}

TEST_F(QDMISessionTest, Allocation) {
  // Checkpoints 1 and 2: Verified by SetUp/TearDown

> }

Reviewer: It bothers me a bit that these two test cases are identical. I understand your motivation for that, and I definitely appreciate the clear failure message that indicates which checkpoint failed. I wonder whether the following scenario would facilitate that every Checkpoint can be associated with one Test case, so that when it succeeds, it also indicates passing the checkpoint.

1. Create yet another test class `QDMIInitializedSessionTest` inheriting from `QDMISessionTest` (I am open for better names here). 
1. Separating the logic in the `SetUp` function of the `QDMISessionTest` class into two parts. One remains, the other is placed in the `SetUp` function of the new class.
1. Adapting the test cases accordingly.

> TEST_F(QDMISessionTest, QueryProperties) {

Reviewer: Please extend the tests such that it covers every failure scenario (as far as possible, MEMOUT might be hard to test for). Also applies to the next test case.

  size_t size = 0;
  ASSERT_EQ(tutorial_QDMI_device_session_query_device_property(session, QDMI_DEVICE_PROPERTY_NAME, 0, nullptr, &size), QDMI_SUCCESS)
      << "Checkpoint 3 Failed: Device failed to report name size.";

  std::string value(size, '\0');
  ASSERT_EQ(tutorial_QDMI_device_session_query_device_property(session, QDMI_DEVICE_PROPERTY_NAME, size, value.data(), nullptr), QDMI_SUCCESS);

>   EXPECT_STREQ(value.c_str(), "MyTutorialDevice");

Reviewer: Imho, the tests do not need to require a certain name for the device. The developer could decide to return another name because they are implementing their own device based on the tutorial. I would still like to have the tests passing in this scenario.

| `QDMI_ERROR_BADSTATE`         | Function called in an incorrect order (e.g., submitting before initialization). |
| `QDMI_ERROR_OUTOFMEM`         | Memory allocation failed.                                                       |

### Debugging Tips

- **Check failure messages**: Test failures include checkpoint messages pointing to the relevant guide section.
- **Detailed output**: Use `ctest --output-on-failure` to see full execution logs.
- **Stub verification**: Ensure you have removed all `QDMI_ERROR_NOTIMPLEMENTED` returns from implemented logic.

## Summary and Next Steps {#tutorial-next-steps}

Congratulations on building your first QDMI device. You've implemented:

1.  **Lifecycle Management**: Handling global and session-level initialization.
2.  **Opaque Data Retrieval**: Mastering the two-step pattern for variable-sized data.
3.  **Job Orchestration**: Moving tasks through a standardized lifecycle.

**What's next?**

- **Hardware Integration**: Replace simulated results with calls to your quantum controller's C/C++ API.
- **Extended Queries**: Add more device properties such as gate fidelities or coupling maps.

> - **Python Integration**: Test your implementation using the provided Python bindings.

Reviewer: There are no real python bindings part of the device at this state. The purpose of the Python package is primarily to provide a way of distributing the compiled dynamic library. I'd suggest to simply remove this point.

---

## Appendix {#tutorial-appendix}

### Configuring the Template

We recommend pinning your project to a specific QDMI release for stability. Edit `cmake/ExternalDependencies.cmake`:

```diff
-   set(QDMI_REV "develop")
+   set(QDMI_REV "v1.3.0")
```

### Troubleshooting Fetch Failures

- **GitHub Rate Limits**: If you encounter 403 errors, set a `GITHUB_TOKEN` environment variable and CMake will use it for authentication.
- **SSL Certificates**: Update your system certificates if the HTTPS fetch fails.
- **Missing Tools**: Ensure `git` and `cmake` are available in your path.

### Build Configuration

- **Debug Configuration**: Use `-DCMAKE_BUILD_TYPE=Debug` during configuration for better stack traces.

> - **Targeted Build**: Build specific components using `--target tutorial-qdmi-device` to save time.

Reviewer: Add a recommendation to install `ccache` to speed up builds as well.
